api/src/test/java/org/openmrs/api/impl/DiagnosisServiceImplTest.java

```diff
@@ -330,5 +343,212 @@ public void purgeDiagnosis_shouldPurgeDiagnosis() {
+		assertEquals("Testing Reference", diagnosisAttribute.getValueReference());
+		assertEquals(1, diagnosisAttribute.getId());
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#getDiagnosisAttributeByUuid(String)
+	 */
+	@Test
+	public void getDiagnosisAttributeByUuid_shouldReturnNullIfNoDiagnosisAttributeHasTheProvidedUuid() {
+		assertNull(diagnosisService.getDiagnosisAttributeByUuid("001612a5-9f2a-4a05-9384-755679f75647"));
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#save(Diagnosis)
+	 */
+	@Test
+	public void saveDiagnosis_shouldSaveTheDiagnosisWithTheProvidedAttributes() {
+		final String NAMESPACE = "namespace";
+		final String FORMFIELD_PATH = "formFieldPath";
+		final DiagnosisAttributeType DIAGNOSIS_ATTRIBUTE_TYPE = Context.getDiagnosisService()
+				.getDiagnosisAttributeTypeByUuid("949daf5b-a83e-4b65-b914-502a553243d3");
+		DiagnosisAttribute diagnosisAttribute = new DiagnosisAttribute();
+		diagnosisAttribute.setAttributeType(DIAGNOSIS_ATTRIBUTE_TYPE);
+		diagnosisAttribute.setCreator(Context.getUserService().getUser(1));
+		diagnosisAttribute.setVoided(false);
+		diagnosisAttribute.setValueReferenceInternal("Diagnosis Attribute Reference");
+		final String UUID = "3c6422d8-7bdd-4e20-bd1b-a590f084db08";
+		Condition condition = conditionService.getConditionByUuid("2cc6880e-2c46-15e4-9038-a6c5e4d22fb7");
+		Encounter encounter = encounterService.getEncounterByUuid("y403fafb-e5e4-42d0-9d11-4f52e89d123r");
+		Patient patient = patientService.getPatient(2);
+		Diagnosis diagnosis = new Diagnosis();
+		diagnosis.setUuid(UUID);
+		diagnosis.setEncounter(encounter);
+		diagnosis.setCondition(condition);
+		diagnosis.setCertainty(ConditionVerificationStatus.CONFIRMED);
+		diagnosis.setPatient(patient);
+		diagnosis.setRank(1);
+		diagnosis.setVoided(false);
+		diagnosis.setFormField(NAMESPACE, FORMFIELD_PATH);
+		diagnosis.addAttribute(diagnosisAttribute);
+		diagnosisService.save(diagnosis);
+		assertNotNull(diagnosis.getId(), "Successfully Saved Diagnosis");
+		Diagnosis savedDiagnosis = diagnosisService.getDiagnosisByUuid(UUID);
+		Set<DiagnosisAttributeType> savedDiagnosisAttributeTypes = savedDiagnosis.getAttributes().stream()
+				.map(DiagnosisAttribute::getAttributeType).collect(Collectors.toSet());
+		assertEquals(condition, savedDiagnosis.getCondition());
+		assertEquals(encounter, savedDiagnosis.getEncounter());
+		assertEquals(patient, savedDiagnosis.getPatient());
+		assertEquals(ConditionVerificationStatus.CONFIRMED, savedDiagnosis.getCertainty());
+		assertEquals(1, savedDiagnosis.getRank());
+		assertEquals(false, savedDiagnosis.getVoided());
+		assertEquals(NAMESPACE + "^" + FORMFIELD_PATH, savedDiagnosis.getFormNamespaceAndPath());
+		assertThat(savedDiagnosis.getAttributes(), hasItem(diagnosisAttribute));
+		assertThat(savedDiagnosisAttributeTypes, contains(DIAGNOSIS_ATTRIBUTE_TYPE));
```

Comment:
Are the last two assertions any different?

Reply:
Yes, one is asserting that the diagnosis has an attribute `assertThat(savedDiagnosis.getAttributes(), hasItem(diagnosisAttribute));` , the last one `assertThat(savedDiagnosisAttributeTypes, contains(DIAGNOSIS_ATTRIBUTE_TYPE));` tests that the attribute has the correct type set at the attributes creation.

Reply:
Fixed